src/cmd/hotspots/add.rs

```diff
@@ -1,106 +1,480 @@
+const COMPRESSION: &str = "cmtDvXumGCrqC1Age74AVPhSRVXJMd8PJS91L8KbNCK";
+const IOT_MINT: &str = "iotEVVZLEywoTn1QdwNPddxPWszn3zFhEot3MfL9fns";
+const DC_MINT: &str = "dcuc8Amr83Wz27ZkQ2K9NS6r8zRpf1J6cvArEBDZDmm";
+
+#[derive(Deserialize, Serialize, Default)]
+struct VerifyResponse {
+    // hex encoded solana transaction
+    pub transaction: String,
+}
+
+#[derive(Deserialize, Serialize, Default)]
+struct VerifyRequest<'a> {
+    // hex encoded solana transaction
+    pub transaction: &'a str,
+    // hex encoded signed message
+    pub msg: &'a str,
+    // hex encoded signature
+    pub signature: &'a str,
+}
 
-    #[structopt(long)]
-    commit: bool,
+#[derive(Debug, Serialize, Deserialize)]
+struct JsonRpcRequest {
+    jsonrpc: String,
+    method: String,
+    params: serde_json::Value,
+    id: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct JsonRpcResponse<T> {
+    jsonrpc: String,
+    result: Option<T>,
+    error: Option<JsonRpcError>,
+    id: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct JsonRpcError {
+    code: i32,
+    message: String,
 }
 
 impl Cmd {
-    pub async fn run(self, opts: Opts) -> Result {
-        let mut txn = BlockchainTxnAddGatewayV1::from_envelope(&read_txn(&self.txn)?)?;
+    pub fn run(&self, opts: Opts) -> Result {
+        let mut add_gateway_txn =
+            BlockchainTxnAddGatewayV1::from_envelope(&read_txn(&self.add_gateway_txn)?)?;
         let password = get_wallet_password(false)?;
-        let wallet = load_wallet(opts.files)?;
+        let wallet = load_wallet(&opts.files)?;
+        let client = new_client(&opts.url)?;
         let keypair = wallet.decrypt(password.as_bytes())?;
-        let staking_client = staking::Client::default();
-        let base_url: String = api_url(wallet.public_key.network);
-        let client = new_client(base_url.clone());
-        let pending_url = base_url + "/pending_transactions/";
-        let wallet_key = keypair.public_key();
-
-        txn.owner_signature = txn.sign(&keypair)?;
-        let envelope = match PublicKey::from_bytes(&txn.payer)? {
-            key if &key == wallet_key => {
-                txn.payer_signature = txn.owner_signature.clone();
-                Ok(txn.in_envelope())
-            }
-            _key if self.onboarding.is_some() && self.commit => {
-                // Only have staking server sign if there's an onboarding key,
-                // and we're actually going to commit
-                let onboarding_key = self.onboarding.as_ref().unwrap().replace('\"', "");
-                staking_client
-                    .sign(&onboarding_key, &txn.in_envelope())
-                    .await
-            }
-            _key => Ok(txn.in_envelope()),
-        }?;
-
-        let status = maybe_submit_txn(self.commit, &client, &envelope).await?;
-        print_txn(&txn, &status, &pending_url, opts.format)
-    }
-}
+        let anchor_client = client.settings.mk_anchor_client(keypair.clone())?;
```

Comment:
not quite the pattern I was hoping for.. why expose the implementation of the request here vs doing it in the `Client`

Reply:
Sure, I can move it into the client. I initially didn't put it there because it's not a function that would be useful elsewhere in the codebase and didn't want to clutter the `Client`